Add implementation for Orders:

DEz-1996 opened the pull request:
1. Add Order.java and OrderItem.java entities
2. Add change files for liquibase
3. Add OrderService.java with OrderServiceImpl.java
4. Add dtos: OrderStatusRequestDto.java, OrderResponseDto.java, OrderItemResponseDto.java, OrderCreateRequestDto.java
5. Add OrderController.java
6. Add mappers: OrderMapper.java and OrderItemMapper.java
7. Add repositories: OrderRepository.java and OrderItemRepository.java

DEz-1996 committed:
Implement Shopping Cart logic:
Add ShoppingCart and CartItem classes;
Add Dto: CartItemAddBookRequestDto, CartItemQuantityRequestDto, ShoppingCartCreateRequestDto, CartItemResponseDto, ShoppingCartResponseDto, ShoppingCartUpdateRequestDto;
Add ShoppingCartMapper and CartItemMapper;
Add changes to create tables "shopping_carts" and "cart_items" in liquibase;
Add ShoppingCartService and ShoppingCartServiceImpl;
Add ShoppingCartRepository and ShoppingCartRepository;
Add ShoppingCartController;

DEz-1996 committed:
Some fixes:
1. Add validation to CartItemAddBookRequestDto and CartItemQuantityRequestDto, add validation in ShoppingCartController methods;
2. Change constraint annotation for field private User user in ShoppingCart
3. Delete redundant ShoppingCartUpdateRequestDto and ShoppingCartCreateRequestDto
4. Some checkstyle fixes

DEz-1996 committed:
Some fixes:
1. Add @Repository to CartItemRepository.java and ShoppingCartRepository.java
2. Change author name in create_shopping_carts_table.yaml and create_cart_items_table.yaml
3. Change name method from getCartContent to getShoppingCart in ShoppingCartController

DEz-1996 committed:
Add implementation for Orders:
1. Add Order.java and OrderItem.java entities
2. Add change files for liquibase
3. Add OrderService.java with OrderServiceImpl.java
4. Add dtos: OrderStatusRequestDto.java, OrderResponseDto.java, OrderItemResponseDto.java, OrderCreateRequestDto.java
5. Add OrderController.java
6. Add mappers: OrderMapper.java and OrderItemMapper.java
7. Add repositories: OrderRepository.java and OrderItemRepository.java

annton-shakur commented on src/main/java/online/book/store/controller/OrderController.java:
```diff
@@ -0,0 +1,84 @@
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/orders")
+@Tag(name = "Orders management", description = "Endpoints for mapping orders")
```
"Endpoints for mapping orders" -> "Endpoints for managing orders"

annton-shakur commented on src/main/java/online/book/store/service/imp/OrderServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
```
I believe that @Transactional annotation is needed here to provide atomicity

annton-shakur commented on src/main/resources/db/changelog/changes/07-create_shopping_carts_table.yaml:
```diff
@@ -0,0 +1,28 @@
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
nullable is redundant here as you have primary key constraint, but that's a minor thing

annton-shakur commented on src/main/resources/db/changelog/changes/08-create_cart_items_table.yaml:
```diff
@@ -0,0 +1,40 @@
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
```
nullable is redundant here as you have primary key constraint, but that's a minor thing

annton-shakur reviewed (approved):
Good job! i have noticed just a few minor things

DEz-1996 committed:
Some fixes:
1. Change swagger description at OrderController
2. Add @Transactional for createOrder method in OrderServiceImpl

DEz-1996 committed:
Merge branch 'main' into HW-15

# Conflicts:
#	src/main/resources/db/changelog/db.changelog-master.yaml

DEz-1996 committed:
Merge branch 'HW-15'

# Conflicts:
#	src/main/resources/db/changelog/db.changelog-master.yaml
rename liquibase change files 11-create_orders_table.yaml from(9-...) and
12-create_order_items_table.yaml from (10-...)

d1sam commented on src/main/java/online/book/store/dto/order/OrderStatusRequestDto.java:
```diff
@@ -0,0 +1,9 @@
+
+@Data
+public class OrderStatusRequestDto {
+    private Order.Status status;
```
```suggestion
    @NotNull
    private Order.Status status;
```

d1sam commented on src/main/java/online/book/store/model/Order.java:
```diff
@@ -0,0 +1,70 @@
+
+    public BigDecimal getTotal() {
+        BigDecimal total = BigDecimal.ZERO;
+        for (OrderItem item : orderItems) {
+            total = total.add(item.getPrice());
+        }
```
can be replaced with Stream API.

d1sam commented on src/main/java/online/book/store/model/Order.java:
```diff
@@ -0,0 +1,70 @@
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
try to avoid EAGER relations.

d1sam commented on src/main/java/online/book/store/model/OrderItem.java:
```diff
@@ -0,0 +1,53 @@
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

d1sam commented on src/main/java/online/book/store/model/OrderItem.java:
```diff
@@ -0,0 +1,53 @@
+    private Order order;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

d1sam commented on src/main/java/online/book/store/repository/order/OrderItemRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+@Repository
+public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {
+    public List<OrderItem> findAllByOrderId(Long orderId);
```
```suggestion
    List<OrderItem> findAllByOrderId(Long orderId, Pageable pageable);
```
don't forget about `Pageable`

d1sam commented on src/main/java/online/book/store/repository/order/OrderRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+@Repository
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    Set<Order> findAllByUserId(Long userId);
```
```suggestion
    List<Order> findAllByUserId(Long userId, Pageable pageable);
```
don't forget about `Pageable`

d1sam commented on src/main/java/online/book/store/service/imp/OrderServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public OrderResponseDto setStatus(OrderStatusRequestDto statusRequestDto, Long orderId) {
```
```suggestion
    @Override
    @Transactional
    public OrderResponseDto updateStatus(OrderStatusRequestDto statusRequestDto, Long orderId) {
```

d1sam commented on src/main/java/online/book/store/service/imp/OrderServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Transactional
+    @Override
+    public OrderResponseDto createOrder(OrderCreateRequestDto createRequestDto, User user) {
```
please divide logic to smaller private methods.

d1sam commented on src/main/java/online/book/store/service/imp/OrderServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+    @Override
+    public Set<OrderItemResponseDto> getAllOrderItems(Long orderId) {
+        List<OrderItem> orderItems = orderItemRepository.findAllByOrderId(orderId);
+        return orderItems.stream()
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toSet());
```
```suggestion
        return orderItemMapper.toDtoList(orderItems);
```
try to use method annotated by `@IterableMapping`

d1sam commented on src/main/java/online/book/store/service/imp/OrderServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+    public Set<OrderResponseDto> getAll(Long userId) {
+        Set<Order> orders = orderRepository.findAllByUserId(userId);
+        return orders.stream()
+                .map(orderMapper::toDto)
```
also use `@IterableMapping`

d1sam reviewed (changes requested):
good job, but there are some changes should be done!

DEz-1996 committed:
Some fixes:
1. Add FetchType.LAZY to fields order and book in OrderItem;
2. Add FetchType.LAZY to field user in Order;
3. Change logic with stream from cycle in Order.getTotal method;
4. Rename orderService.setStatus to orderService.updateStatus;
5. Add pagination to OrderRepository.findAllByUserId and OrderItemRepository.findAllByOrderId;
6. Add @NotNull to field status in OrderStatusRequestDto;
7. Add toDtoList method in OrderMapper and OrderItemMapper;
8. Add pagination to OrderController methods: getAllOrder and getAllOrderItems;
9. Change implementation in OrderServiceImpl.getAllOrderItems and OrderServiceImpl.getAll with pagination;
10. Add @Transactional to OrderServiceImpl.updateStatus method;
11. Split implementation OrderServiceImpl.createOrder method to smaller private methods.
